core/src/main/java/org/infinispan/container/offheap/BoundedOffHeapDataContainer.java

```diff
@@ -209,13 +209,47 @@ private void ensureSize() {
                   addressToRemove = UNSAFE.getLong(firstAddress);
                }
             } else {
-               // This is the only way to break out of loop
+               // This and below are the only ways to break out of entire loop
                break;
             }
          } finally {
             lruLock.unlock();
          }
 
+         // If this was 0 that means we couldn't acquire the write lock with a try so we have to do a regular lock
+         if (addressToRemove == 0) {
+            // Acquire the write lock first and then double check the size before removing
+            entryWriteLock.lock();
+            try {
+               // Now we have to verify the size was greater than max again
+               lruLock.lock();
+               try {
+                  if (currentSize > maxSize) {
+                     // Retrieve the hashCode and make sure it has the same write lock owner
+                     int hashCode = UNSAFE.getInt(firstAddress + 24);
+                     Lock innerLock = locks.getLockFromHashCode(hashCode).writeLock();
+                     if (innerLock == entryWriteLock) {
+                        // If they are the same we can remove the entry
+                        addressToRemove = UNSAFE.getLong(firstAddress);
+                     } else {
+                        // If they weren't the same it means we have a different head now and we have to loop back around
+                        addressToRemove = 0;
+                     }
+                  } else {
+                     // This and above are the only ways to break out of entire loop
+                     break;
```

Comment:
Again, `if (currentSize <= maxSize) break;` looks better to me.

Reply:
Sure

Reply:
Sure